File: backend/data/catalogue.py
from copy import deepcopy
from typing import List

from backend.models import SKU


SEED_CATALOGUE: List[SKU] = [
    SKU(
        id="glowfix-vitamin-c-serum",
        name="GlowFix Vitamin C Serum",
        aliases=[
            "glowfix",
            "glow fix",
            "glowfix serum",
            "vitamin c serum",
            "c serum",
            "serum",
        ],
        base_price_cents=2400,
        current_price_cents=2400,
        stock=42,
        facts=[
            "Brightening serum in a 30 ml bottle",
            "Designed for morning skincare routines",
            "Use before moisturizer and sunscreen",
        ],
    ),
    SKU(
        id="hydramist-cushion-spf",
        name="HydraMist Cushion SPF",
        aliases=[
            "hydramist",
            "hydra mist",
            "hydramist cushion",
            "hydra mist cushion",
            "cushion spf",
            "spf cushion",
            "sunscreen cushion",
            "cushion",
        ],
        base_price_cents=3100,
        current_price_cents=3100,
        stock=28,
        facts=[
            "SPF cushion compact with a dewy finish",
            "Comes in a refillable case",
            "Built for quick touch-ups during the day",
        ],
    ),
    SKU(
        id="bamboo-thermal-tumbler",
        name="Bamboo Thermal Tumbler",
        aliases=[
            "bamboo tumbler",
            "thermal tumbler",
            "bamboo thermal",
            "tumbler",
            "cup",
            "bottle",
        ],
        base_price_cents=1800,
        current_price_cents=1800,
        stock=55,
        facts=[
            "500 ml capacity",
            "Thermal insulation for hot or cold drinks",
            "Bamboo exterior with a reusable design",
        ],
    ),
    SKU(
        id="satin-cloud-sleep-mask",
        name="Satin Cloud Sleep Mask",
        aliases=[
            "satin cloud",
            "sleep mask",
            "satin sleep mask",
            "cloud mask",
            "eye mask",
            "mask",
        ],
        base_price_cents=1200,
        current_price_cents=1200,
        stock=64,
        facts=[
            "Soft satin-feel sleep mask",
            "Adjustable strap for fit",
            "Blocks ambient light during rest",
        ],
    ),
]


def fresh_catalogue() -> List[SKU]:
    return deepcopy(SEED_CATALOGUE)

File: src/lib/catalogue.ts
export type ProductSku = {
  id: string;
  name: string;
  aliases: string[];
  price: string;
  stock: number;
  facts: string[];
};

export const productCatalogue: ProductSku[] = [
  {
    id: "sku-glowfix-vitamin-c-serum",
    name: "GlowFix Vitamin C Serum",
    aliases: ["glowfix", "vitamin c serum", "vit c serum", "glow serum"],
    price: "$42",
    stock: 118,
    facts: [
      "15% vitamin C brightening serum",
      "Fragrance-free lightweight gel texture",
      "Recommended for morning use before sunscreen",
    ],
  },
  {
    id: "sku-hydramist-cushion-spf",
    name: "HydraMist Cushion SPF",
    aliases: ["hydramist", "cushion spf", "spf cushion", "mist cushion"],
    price: "$36",
    stock: 92,
    facts: [
      "SPF 50 cushion compact with dewy finish",
      "Refillable case with mirror and puff",
      "Best for quick touch-ups during the day",
    ],
  },
  {
    id: "sku-bamboo-thermal-tumbler",
    name: "Bamboo Thermal Tumbler",
    aliases: ["bamboo tumbler", "thermal tumbler", "tumbler", "bamboo cup"],
    price: "$28",
    stock: 164,
    facts: [
      "Double-wall insulated stainless steel core",
      "Bamboo-look exterior sleeve",
      "Keeps drinks hot or cold for daily carry",
    ],
  },
  {
    id: "sku-satin-cloud-sleep-mask",
    name: "Satin Cloud Sleep Mask",
    aliases: ["satin mask", "sleep mask", "cloud mask", "satin cloud"],
    price: "$19",
    stock: 246,
    facts: [
      "Soft satin sleep mask with cushioned fill",
      "Adjustable elastic strap",
      "Designed to block light without pressure on eyes",
    ],
  },
];

export const activeSkuId = "sku-glowfix-vitamin-c-serum";

function normalizeSkuText(value: string) {
  return value.trim().toLowerCase().replace(/\s+/g, " ");
}

export function getProductBySkuId(skuId: string) {
  return productCatalogue.find((sku) => sku.id === skuId);
}

export function getProductNameBySkuId(skuId: string) {
  return getProductBySkuId(skuId)?.name;
}

export function resolveSkuFromText(text: string) {
  const normalizedText = normalizeSkuText(text);

  return productCatalogue.find((sku) => {
    const terms = [sku.id, sku.name, ...sku.aliases].map(normalizeSkuText);
    return terms.some((term) => normalizedText.includes(term));
  });
}

export function getActiveSkuDisplay(skuId = activeSkuId) {
  const sku = getProductBySkuId(skuId) ?? productCatalogue[0];

  return {
    id: sku.id,
    name: sku.name,
    aliases: sku.aliases,
    price: sku.price,
    stock: sku.stock,
    facts: sku.facts,
    aliasLabel: sku.aliases[0],
    factSummary: sku.facts.join(" "),
  };
}
